lib/tree.rb

```diff
@@ -1,66 +1,170 @@
-  # Space Complexity: 
+  def print_root(node = @root)
+    if node.nil?
+      puts 'root is nil'
+    else
+      puts node.value
+      puts node.key
+    end
+  end
+
+
+  # Time Complexity:
+  # Space Complexity:
   def add(key, value)
-    raise NotImplementedError
+    current_node = @root
+
+    @root = TreeNode.new(key, value) if current_node.nil?
+
+    until current_node.nil?
+      if (key < current_node.key) && current_node.left.nil?
+        current_node.left = TreeNode.new(key, value)
+      elsif (key > current_node.key) && current_node.right.nil?
+        current_node.right = TreeNode.new(key, value)
+      elsif key < current_node.key
+        current_node = current_node.left
+      elsif key > current_node.key
+        current_node = current_node.right
+      else
+        return
+      end
+    end
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+
+  def find_recursive(current, key)
+    return nil if current.nil?
+
+    return current.value if key == current.key
+
+    if key < current.key
+      find_recursive(current.left, key)
+    elsif key > current.key
+      find_recursive(current.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def find(key)
-    raise NotImplementedError
+    find_recursive(@root, key)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+
+  def inorder_recursion(current, traverse_array)
+    return if current.nil?
+
+    inorder_recursion(current.left, traverse_array)
+
+    traverse_array << { key: current.key, value: current.value }
+
+    inorder_recursion(current.right, traverse_array)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def inorder
-    raise NotImplementedError
+    traverse_array = []
+
+    inorder_recursion(@root, traverse_array)
+
+    traverse_array
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def preorder_recursion(current, traverse_array)
+    return if current.nil?
+
+    traverse_array << { key: current.key, value: current.value }
+    preorder_recursion(current.left, traverse_array)
+    preorder_recursion(current.right, traverse_array)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def preorder
-    raise NotImplementedError
+    traverse_array = []
+
+    preorder_recursion(@root, traverse_array)
+
+    traverse_array
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def postorder_recursion(current, traverse_array)
+    return if current.nil?
+
+    postorder_recursion(current.left, traverse_array)
+    postorder_recursion(current.right, traverse_array)
+    traverse_array << { key: current.key, value: current.value }
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def postorder
-    raise NotImplementedError
+    traverse_array = []
+
+    postorder_recursion(@root, traverse_array)
+
+    traverse_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity:
+  # Space Complexity:
+
   def height
```

Comment:
⚠️  , Time complexity O(n) and space complexity O(h) where `h` is the height of the tree.

However I don't think this method works because it assumes the longest lengths of the trees are on the left and right edges, not somewhere in the middle.

An alternative solution which I think does work

```ruby
    def height_helper(current, count)
      return count if current.nil?

      return [height_helper(current.left, count + 1), height_helper(current.left, count + 1)].max
    end
```